torch/fx/subgraph_rewriter.py

```diff
@@ -0,0 +1,281 @@
+
+        # Terminate early if the node attributes are not equal
+        if not attributes_are_equal(pn, gn):
+            return False
+
+        # Optimistically mark `pn` as a match for `gn`
+        self.nodes_map[pn] = gn
+
+        # Traverse the use-def relationships to ensure that `pn` is a true
+        # match for `gn`
+        if (pn.op != "output"
+                and len(pn.all_input_nodes) != len(gn.all_input_nodes)):
+            return False
+        match_found = all(self._match_nodes(pn_, gn_) for pn_, gn_
+                          in zip(pn.all_input_nodes, gn.all_input_nodes))
+        if not match_found:
+            self.nodes_map.pop(pn)
+            return False
+
+        return True
+
+
+def replace_pattern(gm : GraphModule, pattern : Callable, replacement : Callable) -> None:
+    """
+    Matches all possible non-overlapping sets of operators and their
+    data dependencies (``pattern``) in the Graph of a GraphModule
+    (``gm``), then replaces each of these matched subgraphs with another
+    subgraph (``replacement``).
+
+    Args:
+        ``gm``: The GraphModule that wraps the Graph to operate on
+        ``pattern``: The subgraph to match in ``gm`` for replacement
+        ``replacement``: The subgraph to replace ``pattern`` with
+
+    Examples:
+
+    .. code-block:: python
+
+        import torch
+        from torch.fx import symbolic_trace, subgraph_rewriter
+
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, w1, w2):
+                m1 = torch.cat([w1, w2]).sum()
+                m2 = torch.cat([w1, w2]).sum()
+                return x + torch.max(m1) + torch.max(m2)
+
+        def pattern(w1, w2):
+            return torch.cat([w1, w2]).sum()
+
+        def replacement(w1, w2):
+            return torch.stack([w1, w2])
+
+        traced_module = symbolic_trace(M())
+
+        subgraph_rewriter.replace_pattern(traced_module, pattern, replacement)
+
+    The above code will first match ``pattern`` in the ``forward``
+    method of ``traced_module``. Pattern-matching is done based on
+    use-def relationships, not node names. For example, if you had
+    ``p = torch.cat([a, b])`` in ``pattern``, you could match
+    ``m = torch.cat([a, b])`` in the original ``forward`` function,
+    despite the variable names being different (``p`` vs ``m``).
+
+    The ``return`` statement in ``pattern`` is matched based on its
+    value only; it may or may not match to the ``return`` statement in
+    the larger graph. In other words, the pattern doesn't have to extend
+    to the end of the larger graph.
+
+    When the pattern is matched, it will be removed from the larger
+    function and replaced by ``replacement``. If there are multiple
+    matches for ``pattern`` in the larger function, each non-overlapping
+    match will be replaced. In the case of a match overlap, the first
```

Comment:
This would be first in topological ordering, right? Probably want to mention that, since I don't think the rest of the comment discusses the iteration order of the matching